docs/MATCHMAKING_PROPOSAL.md

```diff
@@ -0,0 +1,171 @@
+This will update the `users_state` map of the game document to `READY` for the player. (The player's uid is the key.)
+
+If after a certain amount of time, the game is not ready, the game is cancelled:
+- The document in the queue is marked as done by setting the `done` field to `true`.
+- The game document is marked as done by setting the `done` field to `true`.
+
+### 5. Round
+Once every player is ready, server adds a round document in the `rounds/` collection of the game document `/games/{gameId}/rounds/{roundId}`:
+```jsonc
+// round.json
+
+{
+  "uid": "round_uid",
+  "timestamp": "2020-01-01T00:00:00.000Z",
+  "players": [/*players*/],
+  "game": "game_uid",
+  "hands": {}
+}
+```
+It also adds the round's `id` to the `rounds` field of the game document.
+
+Each player pick their hand (`ROCK`, `PAPER` or `SCISSORS`). The client update the `hands` map of the round document to reflect the player's hand. Say `user1` picked `ROCK`, `user2` picked `PAPER` and `user3` picked `SCISSORS`:
+
+```jsonc
+// round.json
+
+{
+  "uid": "round_uid",
+  "timestamp": "2020-01-01T00:00:00.000Z",
+  "players": [/*players*/],
+  "game": "game_uid",
+  "hands": {
+    "user1": "ROCK",
+    "user2": "PAPER",
+    "user3": "SCISSORS"
+  }
+}
+```
+
+A round is over when each player in `players` has a value in the `hands` map.
+
+The winner of a round is determined by comparing the hands of the players. The winner is the player with the highest score. It is computed like so:
+
+1. Every player gets a score of 0 at the beginning of the round.
+2. Each player is confronted to every other player.
+    - If the player's hand **wins** against the other player's hand, the player **gets a point**.
+    - If the player's hand **loses** against the other player's hand, the player **loses a point**.
+    - If the player's hand **ties** with the other player's hand, **no points are awarded**.
+
+
+### 6. Next Round
+Once a round is over, the server adds a new round document in the `rounds/` collection of the game document `/games/{gameId}/rounds/{roundId}`:
+
+### 7. Game Over
+Once all rounds are over, the game is over.
+The game document is marked as done by setting the `done` field to `true`.
+
+The winner is the player which won the most rounds.
+
+When a game is set to done, a user can't write to the game document anymore.
+
+A list of a users past games can be thus obtained by querying the `games/` where `players` contains the user's uid and done is `true`:
+```kotlin
+val gamesPlayed = gamesCollection.whereArrayContains("players", user.uid).whereEqualTo("done", true).get()
+```
+## States
+
+Games have 3 implicit states, which are:
+
+State       | Condition
+------------|---------------------------------------------------------
+NOT_STARTED | When not all players are ready
+PLAYING     | When players are all ready, and there's a round present.
+FINISHED    | When the `done` field is set to `true`.
```

Comment:
I'd add an explicit filed state in a game. Too much complexity in this workaround.

Reply:
This isn't a [workaround](https://en.wikipedia.org/wiki/Workaround). That can be added, the only thing is, we always havce to be careful when adding explicit states, as they have to be consistent with the implicit ones, and have a real purpose.